docs/devcontainer/mpi4py-tutorial.ipynb

```diff
@@ -0,0 +1,1549 @@
+      "Hello World from rank 4. total ranks=6. host=dfdb9932ea09 out of /home/app\n",
+      "Hello World from rank 5. total ranks=6. host=dfdb9932ea09 out of /home/app\n",
+      "Stopping engine(s): 1699501339\n",
+      "engine set stopped 1699501339: {'exit_code': 0, 'pid': 35617, 'identifier': 'ipengine-1699501338-ciod-1699501339-35554'}\n",
+      "Stopping controller\n",
+      "Controller stopped: {'exit_code': 0, 'pid': 35587, 'identifier': 'ipcontroller-1699501338-ciod-35554'}\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Hello world first:\n",
+    "import ipyparallel as ipp\n",
+    "\n",
+    "def mpi_example():\n",
+    "    import os\n",
+    "    from mpi4py import MPI\n",
+    "    comm = MPI.COMM_WORLD\n",
+    "    return f\"Hello World from rank {comm.Get_rank()}. total ranks={comm.Get_size()}. host={MPI.Get_processor_name()} out of {os.getcwd()}\"\n",
+    "\n",
+    "# request an MPI cluster with 24 engines\n",
+    "with ipp.Cluster(controller_ip=\"*\", engines=\"mpi\", n=6) as rc:\n",
+    "    # get a broadcast_view on the cluster which is best\n",
+    "    # suited for MPI style computation\n",
+    "    view = rc.broadcast_view()\n",
+    "    # run the mpi_example function on all engines in parallel\n",
+    "    r = view.apply_sync(mpi_example)\n",
+    "    # Retrieve and print the result from the engines\n",
+    "    print(\"\\n\".join(r))\n",
+    "# at this point, the cluster processes have been shutdown"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "/home/app\n",
+      "/home/app\n",
+      "Starting 4 engines with <class 'ipyparallel.cluster.launcher.MPIEngineSetLauncher'>\n",
+      "100%|██████████| 4/4 [00:05<00:00,  1.48s/engine]\n",
+      "Stopping engine(s): 1699501347\n",
+      "engine set stopped 1699501347: {'exit_code': 59, 'pid': 35775, 'identifier': 'ipengine-1699501346-0x7o-1699501347-35554'}\n",
+      "Stopping controller\n",
+      "Controller stopped: {'exit_code': 0, 'pid': 35742, 'identifier': 'ipcontroller-1699501346-0x7o-35554'}\n"
+     ]
+    }
+   ],
+   "source": [
+    "import ipyparallel as ipp\n",
+    "import os\n",
+    "print(f'{os.getcwd()}')\n",
+    "os.chdir('/home/app')\n",
+    "print(f'{os.getcwd()}')\n",
+    "\n",
+    "def run_exago():\n",
+    "    from mpi4py import MPI\n",
+    "    import exago\n",
+    "    comm = MPI.COMM_WORLD\n",
+    "    rank = comm.Get_rank()\n",
+    "    exago.initialize(\"app\", comm)\n",
+    "    opf = exago.OPFLOW()\n",
+    "    opf.read_mat_power_data('./datafiles/case9/case9mod.m')\n",
+    "    opf.solve()\n",
+    "\n",
+    "    comm.barrier()\n",
+    "\n",
+    "    if rank == 0:\n",
+    "        # opf.print_solution()\n",
+    "        opf.save_solution(exago.OutputFormat.JSON, 'solution')\n",
```

Comment:
Suppose I should've looked if there was already a JSON output option... but the csv version has less info than the json and I organized the data slightly differently so I guess it does have a purpose, I'll investigate further on the actual API implementation side